README.md
A standalone library to use the Razer Hydra, based on the findings of 
Ryan Pavlik (https://github.com/rpavlik/razer-hydra-hid-protocol) 
and his code from the VRPN software (http://www.cs.unc.edu/Research/vrpn/)

This was done because there was not a dedicated API for the Hydra, and the Sixense API
was generic for the Sixense products, starting 4 threads as that is the max number of 
controller bases they support (as the Hydra can not change its magnetic frequency, 
only has sense to plug just one). 

I've tried to maintain the code as portable as possible, but I've only tested it on Linux.

Lerna requires the hidapi library.
The test example is done using GLEW and GLFW.


debian/changelog
tinycthread (1.1-1) unstable; urgency=low

  * Initial packaging. (Closes: #)

 -- Klee Dienes <klee@mit.edu>  Thu, 13 Feb 2014 07:53:01 -0500


test/main.cpp
#include <lerna.h>
#include <GL/glew.h>
#include <GL/glfw.h>
#include <cstdio>

bool run = true;
unsigned int winheight = 600, winwidth = 800;

void init3D() {
  glClearColor(0.7f, 0.7f, 0.7f, 1.f);

  glMatrixMode(GL_PROJECTION);
  glLoadIdentity();
  gluPerspective(90.f, 800.f/600.f, 0.1f, 15.f);
  glMatrixMode(GL_MODELVIEW);

  glPointSize(4);
  glLineWidth(4);
  glClearDepth(1.f);
  glDepthFunc(GL_LEQUAL);
  glEnable(GL_DEPTH_TEST);
//glEnable(GL_CULL_FACE);
}

lernaDualControllerData ldcd;
float mat[16]={1.f, 0.f, 0.f, 0.f,
              0.f, 1.f, 0.f, 0.f,
              0.f, 0.f, 1.f, 0.f,
              0.f, 0.f, 0.f, 1.f};
void draw3D() {
  glLoadIdentity();
  glTranslatef(0.f, -1.f, -3.f);

  if(LERNA_OK == lernaGetDualControllerData(&ldcd)) {
    for(int i=0; i<2; i++){
      //Rotation
#define Q ldcd.data[i].quat
      mat[0] = 1.f - 2.f*Q[2]*Q[2] - 2.f*Q[3]*Q[3];
      mat[1] = 2.f*Q[1]*Q[2] + 2.f*Q[0]*Q[3];
      mat[2] = 2.f*Q[1]*Q[3] - 2.f*Q[0]*Q[2];

      mat[4] = 2.f*Q[1]*Q[2] - 2.f*Q[0]*Q[3];
      mat[5] = 1.f - 2.f*Q[1]*Q[1] - 2.f*Q[3]*Q[3];
      mat[6] = 2.f*Q[2]*Q[3] + 2.f*Q[0]*Q[1];

      mat[8] = 2.f*Q[1]*Q[3] + 2.f*Q[0]*Q[2];
      mat[9] = 2.f*Q[2]*Q[3] - 2.f*Q[0]*Q[1];
      mat[10] = 1.f - 2.f*Q[1]*Q[1] - 2.f*Q[2]*Q[2];
#undef Q
      //Position
      mat[12] = ldcd.data[i].pos[0];
      mat[13] = ldcd.data[i].pos[1];
      mat[14] = ldcd.data[i].pos[2];
      glPushMatrix();
      glMultMatrixf((const float*)&mat);
      glBegin(GL_TRIANGLES);
      glColor3f(0.f, 0.f, 1.f);
      glVertex3f(0.1f, 0.f, 0.5f);
      glVertex3f(-0.1f, 0.f, 0.5f);
      glVertex3f(0.f, -0.2f, 0.5f);

      glColor3f(1.f, 0.f, 0.f);
      glVertex3f(0.2f, 0.f, 0.f);
      glColor3f(0.f, 0.f, 1.f);
      glVertex3f(-0.2f, 0.f, 0.f);
      glColor3f(0.f, 1.f, 0.f);
      glVertex3f(0.f, -0.5f, 0.f);
      glEnd();
      glPopMatrix();
    }
  }

  glBegin(GL_LINES);
  glColor3f(1.f, 0.f, 0.f);
  glVertex3f(0.f, 0.f, 0.f);
  glVertex3f(0.5f, 0.f, 0.f);
  glColor3f(0.f, 1.f, 0.f);
  glVertex3f(0.f, 0.f, 0.f);
  glVertex3f(0.f, 0.5f, 0.f);
  glColor3f(0.f, 0.f, 1.f);
  glVertex3f(0.f, 0.f, 0.f);
  glVertex3f(0.f, 0.f, 0.5f);
  glEnd();
}

void GLFWCALL key(int k, int action) {
  if(action!=GLFW_PRESS) return;
  if(k==GLFW_KEY_ESC){
    run = false;
  }
}

void GLFWCALL resize(int width, int height) {
  glMatrixMode(GL_PROJECTION);
  if(height>0) {
    glLoadIdentity();
    gluPerspective(90.f, (float)width/(float)height, 0.1f, 15.f);
  }
  glMatrixMode(GL_MODELVIEW);
}

int main(int argc, char** argv) {
  if(!glfwInit()) return -1;

  glfwOpenWindowHint(GLFW_OPENGL_VERSION_MAJOR, 2);
  glfwOpenWindowHint(GLFW_OPENGL_VERSION_MINOR, 1);

  if(LERNA_OK == lernaInit()){
    if(glfwOpenWindow(800, 600, 0,0,0,0,16,0, GLFW_WINDOW)) {
      glewInit();
      glfwSetKeyCallback(key);
      glfwSetWindowSizeCallback(resize);
      glGetError();

      init3D();
      lernaEnableFiltering(EXP_SMOOTH);
      lernaSetFilterParameter(EXP_SMOOTH, EXP_SMOOTH_ALPHA, 0.8f);

      while(run){
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
        draw3D();

        glfwSwapBuffers();

        unsigned int err = glGetError();
        if(err) printf("%i\n", err);
      }
    }
    lernaExit();
  }

  glfwTerminate();
  return 0;
}


inc/lerna.h
/*
 * Lerna
 * a standalone HID-based API for the Razer Hydra
 * based on the findings of Ryan Pavlik (https://github.com/rpavlik/razer-hydra-hid-protocol)
 * and his code from the VRPN software (http://www.cs.unc.edu/Research/vrpn/)
 *
 * Copyright (c) 2013 Jesus Ojeda
 *
 * Permission is hereby granted, free of charge, to any person or organization
 * obtaining a copy of the software and accompanying documentation covered by 
 * this license (the "Software") to use, reproduce, display, distribute, 
 * execute, and transmit the Software, and to prepare derivative works of the
 * Software, and to permit third-parties to whom the Software is furnished to
 * do so, all subject to the following:
 *
 * The copyright notices in the Software and this entire statement, including
 * the above license grant, this restriction and the following disclaimer, must
 * be included in all copies of the Software, in whole or in part, and all
 * derivative works of the Software, unless such copies or derivative works are
 * solely in the form of machine-executable object code generated by a source
 * language processor.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
 * SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
 * FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
*/

#ifndef _LERNA_H_
#define _LERNA_H_

#ifdef __cplusplus
extern "C" {
#endif

//BUTTON BITMASK
#define BUTTON1 0x04
#define BUTTON2 0x08
#define BUTTON3 0x02
#define BUTTON4 0x10
#define BUT_MID 0x20
#define BUT_JOY 0x40
#define BUMPER  0x01

#define LERNA_OK 0x0
#define LERNA_ERROR 0x1
#define LERNA_HID_ERROR 0x3
#define LERNA_THREADING_ERROR 0x5

/* FILTERS
 * EXP_SMOOTH
 *  - simple interpolation
 *  - Requires ALPHA
 * EXP_SMOOTH_RANGE
 *  - simple interpolation between 0-NEAR (with NALPHA) and FAR-infty (with FALPHA)
 *  - interpolation between NEAR-FAR with NALPHA-FALPHA interpolation
 *  - Requires NEAR, FAR, NALPHA, FALPHA (NEAR, FAR in this order and NEAR < FAR)
 */
typedef enum {EXP_SMOOTH, EXP_SMOOTH_RANGE} filter; //One more filter to come
typedef enum {
  EXP_SMOOTH_ALPHA, 
  EXP_SMOOTH_RANGE_NEAR, 
  EXP_SMOOTH_RANGE_FAR, 
  EXP_SMOOTH_RANGE_NALPHA, 
  EXP_SMOOTH_RANGE_FALPHA
} filter_param;
typedef enum {LEFT=0, RIGHT=1} controller;

typedef struct _controllerdata {
  float pos[3]; //XYZ
  float quat[4]; //WXYZ
  float joy_x, joy_y, trigger;
  unsigned char buttons;
  controller which;
} lernaControllerData;

typedef struct _dual_controllerdata {
  lernaControllerData data[2];
} lernaDualControllerData;


int lernaInit(void);
int lernaExit(void);

int lernaIsActive(void);

int lernaGetDualControllerData(lernaDualControllerData*);
int lernaGetControllerData(controller, lernaControllerData*);

//unsigned char for how far back in history is the data to be retrieved
//max at 64, otherwise LERNA_ERROR is returned
int lernaGetHistoryDualControllerData(unsigned char, lernaDualControllerData*);
int lernaGetHistoryControllerData(unsigned char, controller, lernaControllerData*);

int lernaEnableFiltering(filter);
int lernaDisableFiltering();
int lernaSetFilterParameter(filter, filter_param, float);
int lernaGetFilterParameter(filter, filter_param, float*);

#ifdef __cplusplus
}
#endif

#endif // _LERNA_H_


src/lerna.c
/*
 * Lerna
 *
 * Copyright (c) 2013 Jesus Ojeda
 *
 * Permission is hereby granted, free of charge, to any person or organization
 * obtaining a copy of the software and accompanying documentation covered by 
 * this license (the "Software") to use, reproduce, display, distribute, 
 * execute, and transmit the Software, and to prepare derivative works of the
 * Software, and to permit third-parties to whom the Software is furnished to
 * do so, all subject to the following:
 *
 * The copyright notices in the Software and this entire statement, including
 * the above license grant, this restriction and the following disclaimer, must
 * be included in all copies of the Software, in whole or in part, and all
 * derivative works of the Software, unless such copies or derivative works are
 * solely in the form of machine-executable object code generated by a source
 * language processor.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
 * SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
 * FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
*/

#include "lerna.h"
#include <hidapi/hidapi.h>

#include <tinycthread/tinycthread.h>
// #include <threads.h>
// #include <pthread.h>

#include <stdio.h>
#include <math.h>
#include <string.h>

#define INT16_FLOAT_MM  1e-3f
#define INT16_FLOAT_M11 3.0517578125e-5f // 1.f/32768.f
#define UINT8_FLOAT_01  3.9215686274e-3f // 1.f/255.f

#define HYDRA_VENDORID 0x1532
#define HYDRA_PRODUCTID 0x0300

#define LERNA_SENDING 0x1
#define LERNA_SPHTRAC 0x2
#define LERNA_MAX_HISTORY 65 //Intended MAX + 1
#define LERNA_EPSILON 1e-4f  //Max jump for hemisphere tracking

typedef unsigned char ubyte;
typedef short int16;

const ubyte HYDRA_FEATURE_REPORT[] = {
  0x00, // first byte must be report type
  0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x04, 0x03, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x06, 0x00
};
const int HYDRA_FEATURE_REPORT_LEN = 91;
const ubyte HYDRA_GAMEPAD_COMMAND[] = {
  0x00, // first byte must be report type
  0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x00, 0x00, 0x00, 0x00, 0x05, 0x00
};
const int HYDRA_GAMEPAD_COMMAND_LEN = 91;
const ubyte CD_offset[2] = {8, 30}; //inital offset for each controller_data in buffer

struct _lerna_internaldata {
  ubyte _runTh;
  ubyte _status;
#if 0
  thrd_t _th;
#else
  pthread_t _th;
#endif
  hid_device *_hydra_ctl;
  hid_device *_hydra_dat;
  time_t _time;
  void (*_filter)();
} _lid;

struct _lerna_internal_controller_data {
  float _hemi_mirror[2];
  lernaDualControllerData _history[256];
  ubyte _prev, _last;
} _lcd;

void _hemisphere_tracking(int which_cont) {
#define LAST _lcd._history[_lcd._last].data[which_cont]
#define PREV _lcd._history[_lcd._prev].data[which_cont]
  float old[3] = {PREV.pos[0], PREV.pos[1], PREV.pos[2]};
  float report[3] = {LAST.pos[0], LAST.pos[1], LAST.pos[2]};
  float negat[3] = {-LAST.pos[0], -LAST.pos[1], -LAST.pos[2]};

  float repo = (report[0] - old[0]) * (report[0] - old[0]) +
    (report[1] - old[1]) * (report[1] - old[1]) +
    (report[2] - old[2]) * (report[2] - old[2]);
  float nega = (negat[0] - old[0]) * (negat[0] - old[0]) +
    (negat[1] - old[1]) * (negat[1] - old[1]) +
    (negat[2] - old[2]) * (negat[2] - old[2]);

  if(repo - nega > LERNA_EPSILON) {
    _lcd._hemi_mirror[which_cont] *= -1.f;
    LAST.pos[0] *= -1.f;
    LAST.pos[1] *= -1.f;
    LAST.pos[2] *= -1.f;
  }
#undef LAST
#undef PREV
}

void _quat_normalize(float v[4]) {
  float invlength = 1.f / sqrt(v[0] * v[0] + 
      v[1] * v[1] + 
      v[2] * v[2] +
      v[3] * v[3]);
    v[0] *= invlength;
    v[1] *= invlength;
    v[2] *= invlength;
    v[3] *= invlength;
}

void _processData(ubyte buf[]) {
  lernaDualControllerData *_new = &_lcd._history[buf[7]];
  _lcd._prev = _lcd._last;
  _lcd._last = buf[7];
  int16 *int16buf;
  int i;
  for(i=0; i<2; i++) {
    int16buf = (int16*) &buf[CD_offset[i]];
    //Y and Z are interchanged and Y negated to provide a right hand system (GL)
    _new->data[i].pos[0] = int16buf[0] * INT16_FLOAT_MM * _lcd._hemi_mirror[i];
    _new->data[i].pos[2] = int16buf[1] * INT16_FLOAT_MM * _lcd._hemi_mirror[i];
    _new->data[i].pos[1] = -int16buf[2] * INT16_FLOAT_MM * _lcd._hemi_mirror[i];
    //Similar change for the rotation to provide the right hand system
    //no need to normalize the quaternion, Hydra provides it with error up to 1e-4
    _new->data[i].quat[1] = int16buf[3] * INT16_FLOAT_M11;
    _new->data[i].quat[0] = -int16buf[4] * INT16_FLOAT_M11;
    _new->data[i].quat[2] = int16buf[5] * INT16_FLOAT_M11;
    _new->data[i].quat[3] = int16buf[6] * INT16_FLOAT_M11;
    memcpy(&_new->data[i].buttons, &buf[CD_offset[i]+14], sizeof(ubyte));
    int16buf = (int16*) &buf[CD_offset[i]+15];
    _new->data[i].joy_x = int16buf[0] * INT16_FLOAT_M11;
    _new->data[i].joy_y = int16buf[1] * INT16_FLOAT_M11;
    _new->data[i].trigger = ((float) buf[CD_offset[i]+19]) * UINT8_FLOAT_01;
    _new->data[i].which = i;

    if(_lid._status & LERNA_SPHTRAC) _hemisphere_tracking(i);
    //filter pos and quat
    if(_lid._filter) _lid._filter();
  }
}

void _startStreaming(void) {
  hid_send_feature_report(_lid._hydra_ctl, HYDRA_FEATURE_REPORT, HYDRA_FEATURE_REPORT_LEN);
  ubyte buf[91] = {0};
  hid_get_feature_report(_lid._hydra_ctl, buf, 91);

  //First reading (will take up to 4 seconds aprox for the Hydra to start sending data)
  //Set hemisphere tracking - assume controllers in dock: left controller should have negative x, right controller positive x
  _lid._status |= LERNA_SENDING;
  ubyte rbuf[512];
  int ret=-1;
  while(ret<=0) {
    ret = hid_read(_lid._hydra_dat, rbuf, sizeof(rbuf));
    if(ret>0) {
      _lcd._hemi_mirror[0] = 1.f;
      _lcd._hemi_mirror[1] = 1.f;
      _processData(rbuf);
      if(_lcd._history[_lcd._last].data[0].pos[0] > 0.f) {
        _lcd._hemi_mirror[0] = -1.f;
        _lcd._history[_lcd._last].data[0].pos[0] *= -1.f;
        _lcd._history[_lcd._last].data[0].pos[1] *= -1.f;
        _lcd._history[_lcd._last].data[0].pos[2] *= -1.f;
      }
      if(_lcd._history[_lcd._last].data[1].pos[0] < 0.f) {
        _lcd._hemi_mirror[1] = -1.f;
        _lcd._history[_lcd._last].data[1].pos[0] *= -1.f;
        _lcd._history[_lcd._last].data[1].pos[1] *= -1.f;
        _lcd._history[_lcd._last].data[1].pos[2] *= -1.f;
      }
      _lid._status |= LERNA_SPHTRAC;
    }
    else if(ret<0) fprintf(stderr, "Err\n");
  }

  time(&_lid._time);
}
void _stopStreaming(void) {
  hid_send_feature_report(_lid._hydra_ctl, HYDRA_GAMEPAD_COMMAND, HYDRA_GAMEPAD_COMMAND_LEN);
}

int _hydraProcess(void *arg){
  _startStreaming();

  ubyte rbuf[512];
  int ret;
  while(_lid._runTh) {
    ret = hid_read(_lid._hydra_dat, rbuf, sizeof(rbuf));
    if(ret>0) {
      _processData(rbuf);
    }
    else if(ret<0) fprintf(stderr, "Err\n");
  }
  return 0;
}

int _closehydra() {
  hid_close(_lid._hydra_ctl);
  hid_close(_lid._hydra_dat);
  return (!hid_exit() ? LERNA_OK : LERNA_HID_ERROR);
}

//Acquire Hydra - assume first capable device
int lernaInit(void) {
  _lid._runTh = 0;
  _lid._status = 0;
  if(!hid_init()){
    struct hid_device_info *devs = hid_enumerate(HYDRA_VENDORID, HYDRA_PRODUCTID);
    struct hid_device_info *cur_dev = devs;

    while(cur_dev) { //Expecting only two devices
      if(cur_dev->interface_number) _lid._hydra_ctl = hid_open_path(cur_dev->path);
      else _lid._hydra_dat = hid_open_path(cur_dev->path);

      cur_dev = cur_dev->next;
    }
    hid_free_enumeration(devs);

    if(!_lid._hydra_ctl || !_lid._hydra_dat) {
      hid_exit();
      return LERNA_HID_ERROR;
    }
  }
  if(hid_set_nonblocking(_lid._hydra_ctl, 1) || hid_set_nonblocking(_lid._hydra_dat, 1)){
    _closehydra();
    return LERNA_HID_ERROR;
  }


  _lid._runTh = 1;
  if(thrd_success != thrd_create(&_lid._th, _hydraProcess, NULL)){
    _stopStreaming();
    _closehydra();
    return LERNA_THREADING_ERROR;
  }
  return LERNA_OK;
}

//Release Hydra
int lernaExit(void) {
  if(!_lid._runTh) return LERNA_OK;

  _lid._runTh = 0; //Stop the thread
  int badresult = thrd_success != thrd_join(_lid._th, NULL) ? LERNA_THREADING_ERROR : LERNA_OK;
  _stopStreaming();
  badresult |= _closehydra();
  
  return badresult;
}

//We require 1 second (at 250Hz) to get a full historical controller data buffer (although 1/4 would be enough for a history of 64)
//I known time() has resolution up to seconds, but if we require 1 sec, this function may be portable enough...
int lernaIsActive(void) {
  time_t t2; time(&t2);
  double elapsedTime = difftime(t2, _lid._time);
  return _lid._runTh && (_lid._status & LERNA_SENDING) && elapsedTime>1.0;
}

int lernaGetDualControllerData(lernaDualControllerData *usr) {
  if(_lid._status & LERNA_SENDING)
    memcpy(usr, &_lcd._history[_lcd._last], sizeof(lernaDualControllerData));
  else return LERNA_ERROR;
  return LERNA_OK;
}

int lernaGetControllerData(controller c, lernaControllerData *usr) {
  if(_lid._status & LERNA_SENDING)
    memcpy(usr, &_lcd._history[_lcd._last].data[c], sizeof(lernaControllerData));
  else return LERNA_ERROR;
  return LERNA_OK;
}

int lernaGetHistoryDualControllerData(unsigned char farback, lernaDualControllerData *usr) {
  if(_lid._status & LERNA_SENDING && farback<LERNA_MAX_HISTORY){
    ubyte pos = farback > _lcd._last? 255 - farback + _lcd._last : _lcd._last - farback;
    memcpy(usr, &_lcd._history[pos], sizeof(lernaDualControllerData));
    return LERNA_OK;
  }
  else return LERNA_ERROR;
}

int lernaGetHistoryControllerData(unsigned char farback, controller c, lernaControllerData *usr) {
  if(_lid._status & LERNA_SENDING && farback<LERNA_MAX_HISTORY){
    ubyte pos = farback > _lcd._last? 255 - farback + _lcd._last : _lcd._last - farback;
    memcpy(usr, &_lcd._history[pos].data[c], sizeof(lernaControllerData));
    return LERNA_OK;
  }
  else return LERNA_ERROR;
}

// Filters /////////////////////////////////////////////////////////////////////////////////////////

union _filter_param {
  struct _exp_smooth {
    float alpha;
    float omalpha;
  } _exp_smooth;
  struct _exp_smooth_rng {
    float near, farnear;
    float nalpha, falpha;
  } _exp_smooth_rng;
} _filter_param;

void _lerp_filt(int controller, float alpha, float omalpha) {
#define LAST _lcd._history[_lcd._last].data[controller]
#define PREV _lcd._history[_lcd._prev].data[controller]
  //Position
  LAST.pos[0] = LAST.pos[0] * alpha + PREV.pos[0] * omalpha;
  LAST.pos[1] = LAST.pos[1] * alpha + PREV.pos[1] * omalpha;
  LAST.pos[2] = LAST.pos[2] * alpha + PREV.pos[2] * omalpha;
  //Quaternion (just lerp, assuming similar orientations from contiguous frames)
  LAST.quat[0] = LAST.quat[0] * alpha + PREV.quat[0] * omalpha;
  LAST.quat[1] = LAST.quat[1] * alpha + PREV.quat[1] * omalpha;
  LAST.quat[2] = LAST.quat[2] * alpha + PREV.quat[2] * omalpha;
  LAST.quat[3] = LAST.quat[3] * alpha + PREV.quat[3] * omalpha;

  _quat_normalize(LAST.quat);
#undef LAST
#undef PREV
}

void _filter_Exp_Smooth() {
  _lerp_filt(0, _filter_param._exp_smooth.alpha, _filter_param._exp_smooth.omalpha);
  _lerp_filt(1, _filter_param._exp_smooth.alpha, _filter_param._exp_smooth.omalpha);
}

void _filter_Exp_Smooth_Rng() {
  int i;
  for(i=0; i<2; i++) {
#define LAST _lcd._history[_lcd._last].data[i]
#define ESR _filter_param._exp_smooth_rng
    float t = (LAST.pos[0]*LAST.pos[0] + LAST.pos[1]*LAST.pos[1] + LAST.pos[2]*LAST.pos[2] - ESR.near) / ESR.farnear;
    float tmp_alpha = ESR.nalpha;
    if(t > 0.f) 
      tmp_alpha = t > 1.f ? ESR.falpha : (1.f-t)*ESR.nalpha + t*ESR.nalpha;
    float tmp_omalpha = 1.f - tmp_alpha;

    _lerp_filt(i, tmp_alpha, tmp_omalpha);
#undef LAST
#undef ESR
  }
}

int lernaEnableFiltering(filter fil) {
  switch(fil) {
    case EXP_SMOOTH:
      _lid._filter = &_filter_Exp_Smooth;
      break;
    case EXP_SMOOTH_RANGE:
      _lid._filter = &_filter_Exp_Smooth_Rng;
      break;
    default: 
      return LERNA_ERROR;
  }
  return LERNA_OK;
}

int lernaDisableFiltering() {
  _lid._filter = NULL;
  return LERNA_OK;
}

int lernaSetFilterParameter(filter fil, filter_param fparam, float val) {
  switch(fil) {
    case EXP_SMOOTH:
      if(fparam == EXP_SMOOTH_ALPHA && val >= 0.f && val <= 1.f) {
        _filter_param._exp_smooth.alpha = val;
        _filter_param._exp_smooth.omalpha = 1.f - val;
      } else return LERNA_ERROR;
      break;
    case EXP_SMOOTH_RANGE:
      if(val < 0.f) return LERNA_ERROR;
      switch(fparam) {
        case EXP_SMOOTH_RANGE_NEAR:
          _filter_param._exp_smooth_rng.near = val*val;
          break;
        case EXP_SMOOTH_RANGE_FAR:
            _filter_param._exp_smooth_rng.farnear = val*val - _filter_param._exp_smooth_rng.near;
          break;
        case EXP_SMOOTH_RANGE_NALPHA:
          _filter_param._exp_smooth_rng.nalpha = val;
          break;
        case EXP_SMOOTH_RANGE_FALPHA:
          _filter_param._exp_smooth_rng.falpha = val;
          break;
        default:
          return LERNA_ERROR;
      } break;
    default:
      return LERNA_ERROR;
  }
  return LERNA_OK;
}

int lernaGetFilterParameter(filter fil, filter_param fparam, float *val) {
  if(!val) return LERNA_ERROR;
  switch(fil) {
    case EXP_SMOOTH:
      if(fparam == EXP_SMOOTH_ALPHA)
        *val = _filter_param._exp_smooth.alpha;
      else return LERNA_ERROR;
      break;
    case EXP_SMOOTH_RANGE:
      switch(fparam) {
        case EXP_SMOOTH_RANGE_NEAR:
          *val = sqrt(_filter_param._exp_smooth_rng.near);
          break;
        case EXP_SMOOTH_RANGE_FAR:
          *val = sqrt(_filter_param._exp_smooth_rng.farnear + _filter_param._exp_smooth_rng.near);
          break;
        case EXP_SMOOTH_RANGE_NALPHA:
          *val = _filter_param._exp_smooth_rng.nalpha;
          break;
        case EXP_SMOOTH_RANGE_FALPHA:
          *val = _filter_param._exp_smooth_rng.falpha;
          break;
        default:
          return LERNA_ERROR;
      } break;
    default:
      return LERNA_ERROR;
  }
  return LERNA_OK;
}

